consensus/polybft/polybft.go

```diff
@@ -0,0 +1,662 @@
+				p.logger.Warn(fmt.Sprintf("Failed to deliver message. Error: %s", err))
+			}
+		})
+		if err != nil {
+			return fmt.Errorf("topic subscription failed:%w", err)
+		}
+	}
+
+	return nil
+}
+
+func (p *Polybft) startPbftProcess() {
+	// wait to have at least n peers connected. The 2 is just an initial heuristic value
+	// Most likely we will parametrize this in the future.
+	if !p.waitForNPeers() {
+		return
+	}
+
+	// subscribe to new block events
+	var (
+		newBlockSub   = p.blockchain.SubscribeEvents()
+		syncerBlockCh = make(chan uint64)
+	)
+
+	// Receive a notification every time syncer manages
+	// to insert a valid block.
+	go func() {
+		eventCh := newBlockSub.GetEventCh()
+
+		for {
+			select {
+			case ev := <-eventCh:
+				currentBlockNum := p.blockchain.CurrentHeader().Number
+				if ev.Source == "syncer" {
+					if ev.NewChain[0].Number < currentBlockNum {
+						continue
+					}
+				}
+
+				if p.isSynced() {
+					syncerBlockCh <- currentBlockNum
+				}
+
+			case <-p.closeCh:
+				return
+			}
+		}
+	}()
+
+	defer newBlockSub.Close()
+
+SYNC:
+	if !p.isSynced() {
+		<-syncerBlockCh
+	}
+
+	lastBlock := p.blockchain.CurrentHeader()
+	p.logger.Info("startPbftProcess",
+		"header hash", lastBlock.Hash,
+		"computed hash", lastBlock.Hash,
+		"header number", lastBlock.Number)
+
+	currentValidators, err := p.GetValidators(lastBlock.Number, nil)
+	if err != nil {
+		p.logger.Error("failed to query current validator set", "block number", lastBlock.Number, "error", err)
+	}
+
+	p.runtime.setIsActiveValidator(currentValidators.ContainsNodeID(p.key.NodeID()))
+
+	if !p.runtime.isActiveValidator() {
+		// inactive validator is not part of the consensus protocol and it should just perform syncing
+		goto SYNC
+	}
+
+	// we have to start the bridge snapshot when we have finished syncing
+	if err := p.runtime.restartEpoch(lastBlock); err != nil {
+		p.logger.Error("failed to restart epoch", "error", err)
+
+		goto SYNC
+	}
+
+	for {
+		if err := p.runCycle(); err != nil {
+			if errors.Is(err, errNotAValidator) {
+				p.logger.Info("Node is no longer in validator set")
+			} else {
+				p.logger.Error("an error occurred while running a state machine cycle.", "error", err)
+			}
+
+			goto SYNC
+		}
+
+		switch p.pbft.GetState() {
+		case pbft.SyncState:
+			// we need to go back to sync
+			goto SYNC
+		case pbft.DoneState:
+			// everything worked, move to the next iteration
+		default:
+			// stopped
+			return
+		}
+	}
+}
+
+// isSynced return true if the current header from the local storage corresponds to the highest block of syncer
+func (p *Polybft) isSynced() bool {
+	// TODO: Check could we change following condition to this:
+	// p.syncer.GetSyncProgression().CurrentBlock >= p.syncer.GetSyncProgression().HighestBlock
+	syncProgression := p.syncer.GetSyncProgression()
+
+	return syncProgression == nil ||
+		p.blockchain.CurrentHeader().Number >= syncProgression.HighestBlock
+}
```

Comment:
I don't think you can rely on sync progression, because the syncing mechanism is always on and always syncing blocks

cc @dbrajovic

Reply:
This was a requirement for the integration with `pbft-consensus`. It is expected that the sync protocol is always syncing blocks.